### PR TITLE
Add modal for Sharing

### PR DESCRIPTION
**Pull request by ramonrovirosa**

### CODE REVIEW

File: app/javascript/share_images/share.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
+$(document).ajaxSuccess((event, xhr) => {
```
Comment (lsq921224) [inline app/javascript/share_images/share.js]: attaching an event listener to `document` is not very good, because we may have multiple ajax calls on the page, and this will be triggered for every kind of ajax calls. What we could do is to pass the selector from the html template, and use it in our javascript file. e.g: 
```
class ShareView {
  constructor(modalSelector) {
    this.$modal= $(modalSelector);
    this.attachEventHandlers();
  }

    attachEventHandlers() {
         this.$modal.ajaxSuccess((event, data) => {
           ...
      })
   }
}
```

File: app/controllers/shares_controller.rb [hunk 2]
```diff
@@ -7,16 +7,29 @@ def show
   end
 
   def create
-    share_image_form = Forms::ShareForm.new(share_params)
-    if share_image_form.valid?
+    share_form = Forms::ShareForm.new(share_params)
+    if share_form.valid?
       send_share_email(share_params)
-      return redirect_with_flash_message(:success, 'Your image was shared.', image_path(@image))
+      return redirect_with_flash_message(:success, 'Your image was shared.', images_path)
```
Comment (lsq921224) [inline app/controllers/shares_controller.rb]: This is not how we usually deal with a XHR request. What we should really do here is like what you did in line 16-20, but instead sending the new form, we send the new path and the flash message, and use javascript on the client side to deal with it

File: app/javascript/share_images/share.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  event.preventDefault();
+  document.write(xhr.responseText);
```
Comment (lsq921224) [inline app/javascript/share_images/share.js]: Per my comment on `shares_controller.rb`, we need to update here as well

File: app/javascript/share_images/share.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  generateModal(originalImageUrl, originalImageId);
+});
+
+// see https://css-tricks.com/snippets/javascript/loop-queryselectorall-matches/
+if(!NodeList.prototype.forEach){
+  NodeList.prototype.forEach = Array.prototype.forEach;
+}
+
+document.querySelectorAll('.modal-btn').forEach(elm => elm.addEventListener('click', (event) => {
+  const {imgUrl, imgId} = event.target.dataset;
```
Comment (lsq921224) [inline app/javascript/share_images/share.js]: If you watch the video attached to this issue https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project/basics-of-bootstrap-modals-js-and-ajax, you will know that there is an event called `show.bs.modal`, so instead of attaching events to all the buttons, we can simply listen to `show.bs.model`

File: app/views/images/show.html.erb [hunk 5]
```diff
@@ -18,8 +18,14 @@
...
-  <%= form.submit 'Delete', class: 'btn btn-primary homepage-button' %>
+  <%= form.button 'Delete', class: 'btn btn-primary homepage-button margin-bottom-20' %>
 <% end %>
+
+<button type="button" class="btn btn-primary modal-btn" data-toggle="modal" data-target="#shareModal" data-img-url="<%= @image.url %>" data-img-id="<%= @image.id %>">
+  Share
+</button>
+
+<%= render partial: "shared/share_modal", locals: { share_form: @share_form } %>
+
```
Comment (lsq921224) [inline app/views/images/show.html.erb]: Can we remove this empty line? In APM you will receive comments from `farcy`, our code style robot

File: app/views/layouts/application.html.erb [hunk 6]
```diff
@@ -18,12 +18,14 @@
     <%= render partial: "shared/nav_header" %>
     <%= yield %>
 
+    <%= javascript_include_tag 'application' %>
     <!-- Optional JavaScript -->
     <!-- jQuery first, then Popper.js, then Bootstrap JS -->
-    <script src="https://code.jquery.com/jquery-3.2.1.slim.min.js" integrity="sha384-KJ3o2DKtIkvYIK3UENzmM7KCkRr/rE9/Qpg6aAZGJwFDMVNA/GpGFF93hXpG5KkN" crossorigin="anonymous"></script>
+    <!--<script src="https://code.jquery.com/jquery-3.3.1.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>-->
```
Comment (lsq921224) [inline app/views/layouts/application.html.erb]: Is this still needed?

File: app/views/layouts/application.html.erb [hunk 7]
```diff
@@ -18,12 +18,14 @@
...
     <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.12.3/umd/popper.min.js" integrity="sha384-vFJXuSJphROIrBnz7yo7oB41mKfc8JzQZiCq4NCceLEaO4IHwicKwpJf9c9IpFgh" crossorigin="anonymous"></script>
     <script src="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0-beta.2/js/bootstrap.min.js" integrity="sha384-alpBpkh1PFOepccYVYDB4do5UnbKysX5WZXm3XxPqe5iKTfUKjNkCk9SaVuEZflJ" crossorigin="anonymous"></script>
 
     <%= javascript_pack_tag 'images' %>
+    <%= javascript_pack_tag 'share' %>
```
Comment (lsq921224) [inline app/views/layouts/application.html.erb]: I don't think we should include this line here, the previous line should not be here either, because this is `application`'s layout. Imagine if we were to create a users page, then `images` and `share` javascripts will be completely useless..
so a better way here is to `<%= yield :javascript %>` here, and in our `images` and `share` html template, we do 
```
<% content_for :javascript do %>
  <%= javascript_pack_tag '...‘ %>
<% end %>
```

File: app/views/shared/_share_modal.html.erb [hunk 8]
```diff
@@ -0,0 +1,27 @@
+<!-- Modal -->
+<div class="modal fade" id="shareModal" tabindex="-1" role="dialog" aria-labelledby="exampleModalLabel" aria-hidden="true" data-backdrop="false">
```
Comment (lsq921224) [inline app/views/shared/_share_modal.html.erb]: Can we change `aria-labelledby` to make more sense?

File: app/views/shared/_share_modal.html.erb [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+  <div class="modal-dialog" role="document">
+    <div class="modal-content">
+      <div class="modal-header">
+        <h5 class="modal-title" id="exampleModalLabel">Share Your Image</h5>
```
Comment (lsq921224) [inline app/views/shared/_share_modal.html.erb]: Can we change `id` to make more sense?

File: test/controllers/images_controller_test.rb [hunk 10]
```diff
@@ -80,7 +80,7 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
       post images_path, params: image_params
     end
 
-    assert_response :success
+    assert_response :unprocessable_entity
```
Comment (lsq921224) [inline test/controllers/images_controller_test.rb]: Per previous comment, this should belong to another commit

File: app/controllers/application_controller.rb [hunk 11]
```diff
@@ -1,5 +1,3 @@
 class ApplicationController < ActionController::Base
-  protect_from_forgery with: :exception
```
[the remark this answers is not in the document]
Comment (ramonrovirosa) [reply app/controllers/application_controller.rb]: Addressed here: https://github.com/AppFolioOnboarding/image-sharer-ramonrovirosa/pull/23/commits/47bf505d62c86ef9e0891769e97405589a1b5e25
